text/0006-sampling.md

```diff
@@ -0,0 +1,302 @@
+The sampling decision will happen before a real `Span` object is returned to the user, because:
+ * If child spans are created they need to know the 'SampledFlag'.
+ * If `SpanContext` is propagated on the wire the 'SampledFlag' needs to be set.
+ * If user records any tracing event the `Span` object needs to know if the data are kept or not.
+ It may be possible to always collect all the events until the sampling decision is made but this is
+ an important optimization.
+
+There are two important use-cases to be considered:
+ * All information that may be used for sampling decisions are available at the moment when the 
+ logical `Span` operation should start. This is the most common case.
+ * Some information that may be used for sampling decision are NOT available at the moment when the
+ logical `Span` operation should start (e.g. `http.route` may be determine later).
+
+The current [span creation logic][span-creation] facilitates very well the first use-case, but
+the second use-case requires users to record the logical `start_time` and collect all the
+information necessarily to start the `Span` in custom objects, then when all the properties are
+available call the span creation API.
+
+The RFC proposes that:
+ * For languages where a `Builder` pattern is used to construct a `Span`, to allow users to create a
+ `Builder` where the start time of the Span is considered when the `Builder` is created.
+ * For languages where no intermediate object is used to construct a `Span`, to allow users maybe
+  via a `StartSpanOption` object to start a `Span`. The `StartSpanOption` allows users to set all
+  the start `Span` properties.
+
+**Alternatives considerations:**
+ * We considered, instead of requiring that sampling decision happens before the `Span` is 
+ created to add an explicit `MakeSamplingDecision(SamplingHint)` on the `Span`. Attempts to create
+ a child `Span`, or to access the `SpanContext` would fail if `MakeSamplingDecision()` had not yet
+ been run.
+   * Pros:
+     * Simplifies the case when all the attributes that may be used for sampling are not available
+     when the logical `Span` operation should start.
+   * Cons:
+     * The most common case would have required an extra API call.
+     * Error prone, users may forget to call the extra API.
+     * Unexpected and hard to find errors if user tries to create a child `Span` before calling
+     MakeSamplingDecision().
+ * We considered allowing the sampling decision to be arbitrarily delayed, but guaranteed before 
+ any child `Span` is created, or `SpanContext` is accessed, or before `Span.end()` finished.
+   * Pros:
+     * Similar and smaller API that supports both use-cases defined ahead.
+   * Cons:
+     * If `SamplingHint` needs to also be delayed recorded then an extra API on Span is required 
+     to set this.
+     * Does not allow optimization to not record tracing events, all tracing events MUST be 
+     recorded before the sampling decision is made.
```

Comment:
if sampling decisions are based on a set of policies, it can be an inherent attribute of a policy whether it depends on a delayed datum (such as late tag or span name) and thus requires recording events prior to sampling decision being made. If user picks a policy that does not depend on such delayed data (e.g. user picks a basic probabilistic policy), then the tracer would know that no delayed sampling is going to happen and can short-circuit event recording once the decision is No. This is the approach we're taking in Jaeger SDKs.

Reply:
We deliberately chose to defer the "make decision later" option until later, as it adds more complexity and we can always add it later.

Reply:
We need to define when is the limit when the sampling decision should happen. I like the idea to have on the sampler a property that says is delayed sampling so we keep all the events until the sampling happens, but when is the limit/point that guarantees that sampling happened.

Reply:
see brain dump above https://github.com/open-telemetry/rfcs/pull/24#discussion_r315999705

Reply:
This is in a "alternatives" section and we actually reverted the proposed change. So I would keep it here just for the record and we will discuss in more details when we do make a decision about this.